File: 572-subtree-of-another-tree/572-subtree-of-another-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        # dfs on the tree to find the subroot
        # then perform another dfs to see if same tree
        
        def isSameTree(a,b):
            if not a and not b:
                return True
            elif not a or not b or a.val != b.val:
                return False
            
            return isSameTree(a.left, b.left) and isSameTree(a.right, b.right)
        
        stack = [root]
        while len(stack) > 0:
            node = stack.pop()
            
            if node.val == subRoot.val and isSameTree(node,subRoot):
                return True
            
            if node:
                if node.left:
                    stack.append(node.left)
                if node.right:
                    stack.append(node.right)
                    
            
        return False
        

File: 572-subtree-of-another-tree/572-subtree-of-another-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} subRoot
 * @return {boolean}
 */

// time: O(n^2), n= num of nodes in the tree
// space : O(1), cuz stack always only have 2 items remaining to process
const isSameSubTree = function(a,b) {
    if (a == null && b == null) {
        return true
    } else if (a == null || b == null || a.val !== b.val) {
        return false
    } 
    
    return isSameSubTree(a.left,b.left) && isSameSubTree(a.right, b.right)
}

var isSubtree = function(root, subRoot) {
    
    const node = root
    const stack = [root]
    // 
    while (stack.length > 0) {
        console.log(stack)
        const popped = stack.pop()
        if (popped.val === subRoot.val) {
            // run dfs
            const result = isSameSubTree(popped, subRoot)
            if (result) {
                return true
            }
        }
        
        if (popped.left) {
            stack.push(popped.left)
        }
        
        if (popped.right) {
            stack.push(popped.right)
        }
        
    }
    
    return false
};